### ISSUE TITLE
[Feat] Add Support for onMouseOver and onMouseLeave props in `<AdvancedMarker />`

### ISSUE THREAD

**Issue by rileymiller** [visgl/react-google-maps#304]
### Target Use Case

Adding these props would enable developers to build functionality around AdvancedMarker hover states. e.g I'd like to be able to show a preview of some metadata in an InfoWindow when a marker is being hovered over.

### Proposal

Add `onMouseOver` and `onMouseLeave` as props, similar to the current api of `onClick` and the drag event handlers.

**Comment by rvanderfeer** [comment]
I ran into this too. You can work around this by adding the props to your AdvancedMarker content, for example;

```tsx
<AdvancedMarker
	ref={markerRef}
	position={markerPosition}
	onClick={() => setInfowindowOpen(true)}
	{...other}
>
	<div
		onMouseOver={() => setInfowindowOpen(true)}
		onMouseOut={() => setInfowindowOpen(false)}
		style={{
			width: 20,
			height: 20,
			position: 'absolute',
			top: 0,
			left: 0,
			background: '#F00',
			border: `2px solid #FFF`,
			borderRadius: '50%',
			transform: 'translate(-50%, -50%)'
		}}
	/>
	{infowindowOpen && (
		<InfoWindow
			anchor={marker}
			maxWidth={200}
			onCloseClick={() => setInfowindowOpen(false)}
		>
			infowindow content
		</InfoWindow>
	)}
</AdvancedMarker>
```

**Comment by usefulthink** [comment]
Unfortunately, that doesn't work if your content is just a `Pin` component. I should be able to look into this tomorrow.

**Comment by BParvatkar** [comment]
Were you able to look into it @usefulthink. I was looking into the API reference for AdvanceMarker [here](https://developers.google.com/maps/documentation/javascript/reference/advanced-markers#AdvancedMarkerElement.addEventListener). Seems that the addEventListener method is part of beta version as of now. The APIProvider component does allow us to pass version as prop. Was wondering if this will work:
- Pass version as beta in APIProvider prop
- The AdvancedMarker component extends for all [EventListener](https://developer.mozilla.org/docs/Web/API/EventTarget/addEventListener#listener). 
- The user passes any listener as prop, let say onMouseEnter
- If APIProvider is initialized in beta mode then we allow this events, else give error saying that its still in beta and needs to initialize APIProvider in beta version.

The benefit of this approach is that we don't have to manually add each listener to the components like `AdvancedMarker` or `Pin`

**Comment by usefulthink** [comment]
We're in a bit of a transition-period right now – the version currently in beta will introduce the web-components (`<gmp-map>` and `<gmp-advanced-marker>`), and with that we'll probably also get support for DOM events on the `AdvancedMarkerElement` itself.

But I wouldn't want to make the beta-channel a requirement for using AdvancedMarker, if we can solve this in another way.

Until the web-components are in GA (weekly and quarterly channels), the best thing to do is probably to add the event-listeners to the [`advancedMarker.element`](https://developers.google.com/maps/documentation/javascript/reference/advanced-markers#AdvancedMarkerElement.element) instead. We probably don't even need a different handling of this for when web-components are released, but that remains to be seen.

**Comment by BParvatkar** [comment]
Is there any example of adding listeners to the `advancedMarker.element` or does it need to be built?

**Comment by usefulthink** [comment]
No, it's on my ToDo-list, but I haven't gotten to it yet. I'm still pondering if we should use wrapped events for the markers [like we do for the map](https://github.com/visgl/react-google-maps/blob/f38d3c4004663fd1850c00dca7ddfb7e92b8d5cf/src/components/map/use-map-events.ts#L51-L73) itself, or if we should just use the native events.

**Comment by BParvatkar** [comment]
Earlier it made sense to add each declaration because the type of events was `MapMouseEvent` and not the native events emitted by the element. But as per the docs, they will be emitting native events so I think it will be better to go the native way.

**Comment by usefulthink** [comment]
Yeah, I'm thinking the same. For the map it was necessary because the map-events mostly don't contain any useful data about the event, but here we have perfectly functional DOM-events already :)

**Comment by m-nathani** [comment]
Facing the same issue. `AdvanceMarker` support  `onMouseOver` and  `onMouseEnter` , but unfortunately the events for `onMouseOut` and `onMouseLeave` are not firing... 

kinda dependent on these events to perform some UI states, does anyone found a workaround or react-google-maps have a solution for it yet ?

**Comment by m-nathani** [comment]
Hi @usefulthink ,

I am trying attach event ot AdvanceMarkerElement like as mentioend above to attach event manually:

```ts
const [markerRef, marker] = useAdvancedMarkerRef();

...

useEffect(() => {
  if (marker) {
    marker.addEventListener('mouseleave', handleUnHoverMarker, false);
  }

  return () => {
    if (marker) {
      marker?.removeEventListener('mouseleave', handleUnHoverMarker, false);
     }
  };
}, [marker, handleUnHoverMarker]);

return (
  <>
    <AdvancedMarker
      ref={refCallback}
      onClick={handleOpen}
      position={point}
      title={point.name}
      zIndex={50}
      {...props}
    >

...
```

However, its throwing an error as attached in screenshot. Do you know the version in which it will be available or any work around for this please ? 

![image](https://github.com/visgl/react-google-maps/assets/13331386/05d5865e-8187-436b-9f2f-4c915afabcce)

**Comment by jogoley** [comment]
Did anyone ever figure this out, I am following generally what you are saying but don't know how to implement the addition of listener.

**Comment by jogoley** [comment]
> I ran into this too. You can work around this by adding the props to your AdvancedMarker content, for example;
> 
> ```tsx
> <AdvancedMarker
> 	ref={markerRef}
> 	position={markerPosition}
> 	onClick={() => setInfowindowOpen(true)}
> 	{...other}
> >
> 	<div
> 		onMouseOver={() => setInfowindowOpen(true)}
> 		onMouseOut={() => setInfowindowOpen(false)}
> 		style={{
> 			width: 20,
> 			height: 20,
> 			position: 'absolute',
> 			top: 0,
> 			left: 0,
> 			background: '#F00',
> 			border: `2px solid #FFF`,
> 			borderRadius: '50%',
> 			transform: 'translate(-50%, -50%)'
> 		}}
> 	/>
> 	{infowindowOpen && (
> 		<InfoWindow
> 			anchor={marker}
> 			maxWidth={200}
> 			onCloseClick={() => setInfowindowOpen(false)}
> 		>
> 			infowindow content
> 		</InfoWindow>
> 	)}
> </AdvancedMarker>
> ```

I was able to use this method but was only able to use it after setting version='beta' in the props for APIProvider

**Comment by ewarrenG** [comment]
> Were you able to look into it @usefulthink. I was looking into the API reference for AdvanceMarker [here](https://developers.google.com/maps/documentation/javascript/reference/advanced-markers#AdvancedMarkerElement.addEventListener). Seems that the addEventListener method is part of beta version as of now. The APIProvider component does allow us to pass version as prop. Was wondering if this will work:
> 
> * Pass version as beta in APIProvider prop
> * The AdvancedMarker component extends for all [EventListener](https://developer.mozilla.org/docs/Web/API/EventTarget/addEventListener#listener).
> * The user passes any listener as prop, let say onMouseEnter
> * If APIProvider is initialized in beta mode then we allow this events, else give error saying that its still in beta and needs to initialize APIProvider in beta version.
> 
> The benefit of this approach is that we don't have to manually add each listener to the components like `AdvancedMarker` or `Pin`

Can you share a code example of this? I'm unable to make it work how you describe
